**Betting Edge**
**Exploratory Data Analysis**

In [74]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


In [75]:
excelSheet = pd.ExcelFile('FantasyFootballWeekly.xlsx')
#reading in excel file to 'excelSheet'

print(excelSheet.sheet_names)
#printing sheet names

['FF_Week4_edit.csv', 'FF_Week1_edit.csv', 'FF_Week3_edit.csv', 'FF_Week5_edit.csv', 'FF_Week11_edit.csv', 'FF_Week13_edit.csv', 'FF_Week9_edit.csv', 'FF_Week14_edit.csv', 'FF_Week18_edit.csv', 'FF_Week10_edit.csv', 'FF_Week12_edit.csv', 'FF_Week16_edit.csv', 'FF_Week17_edit.csv', 'FF_Week2_edit.csv', 'FF_Week15_edit.csv', 'FF_Week8_edit.csv', 'FF_Week7_edit.csv', 'FF_Week6_edit.csv']


In [76]:
dataframes = []
#list storing each sheet's dataframe

for sheet_name in excelSheet.sheet_names:
    # Read the sheet into a DataFrame
    weekly_data = excelSheet.parse(sheet_name)

    week_number = int(sheet_name.split('_')[1][4:])
    #extracting the number after "Week"

    weekly_data['Week'] = week_number
    #Adding column for week
    
    # Append the DataFrame to the list
    dataframes.append(weekly_data)

#looping through each sheet name, reading it as a DataFrame, and extracting the week number

dfAll = pd.concat(dataframes, ignore_index=True)
#concatenating all DataFrames into a single DataFrame called 'dfAll'

In [77]:
dfAll.head(5)
#previewing new dataframe to ensure that everything has ran correctly

Unnamed: 0          PLAYER NAME PLAYER TEAM PLAYER POSITION   STATUS  PROJ  \
0           0  Christian McCaffrey          SF              RB  W 35-16  21.1   
1           1           A.J. Brown         Phi              WR  W 34-31  15.7   
2           2           Josh Allen         Buf              QB  W 48-20  20.9   
3           3         Stefon Diggs         Buf              WR  W 48-20  17.8   
4           4         Nico Collins         Hou              WR   W 30-6  12.2   

   PASSING YDS  PASSING TD  PASSING INT  RUSHING CAR  ...  MISC FUML  MISC TD  \
0          0.0         0.0          0.0         20.0  ...        0.0      0.0   
1          0.0         0.0          0.0          0.0  ...        0.0      0.0   
2        320.0         4.0          0.0          4.0  ...        0.0      0.0   
3          0.0         0.0          0.0          0.0  ...        0.0      0.0   
4          0.0         0.0          0.0          0.0  ...        0.0      0.0   

   TOTAL  PASSING COMPLETION  PASSING ATTEMPTS  Opponent  Location  rank  \
0   48.7                 0.0               0.0       Ari      Home   1.0   
1   38.5                 0.0               0.0       Wsh      Home   1.0   
2   36.5                21.0              25.0       Mia      Home   1.0   
3   36.0                 0.0               0.0       Mia      Home   2.0   
4   35.8                 0.0               0.0       Pit      Home   3.0   

       DATE  Week  
0  10-03-23     4  
1  10-03-23     4  
2  10-03-23     4  
3  10-03-23     4  
4  10-03-23     4  

[5 rows x 27 columns]

- After reviewing the data we can see it contains performance metrics for certain offensive positions in the 2023 NFL season 
    - Only certain positions are used in fantasy, due to the fact that certain position's performance are difficult to measure through statistics
    - The only positions used in the vast majority of fantasy football leagues are quarterback, runningback, wide receiver, and tight end
- The data includes player names, teams, positions, game outcomes, projected points, and key stats in passing, rushing, and miscellaneous categories
- The data is structured to analyze player performance for specific weeks, helping with player comparisons, tracking projections vs. actuals, and identifying top performers

In [78]:
dfAll.shape

(18090, 27)

- Checking the shape of the dataframe
    - 18,090 rows
    - 26 columns

In [105]:
dfAll.head()

PLAYER NAME PLAYER TEAM PLAYER POSITION   STATUS  PROJ  \
0  Christian McCaffrey          SF              RB  W 35-16  21.1   
1           A.J. Brown         Phi              WR  W 34-31  15.7   
2           Josh Allen         Buf              QB  W 48-20  20.9   
3         Stefon Diggs         Buf              WR  W 48-20  17.8   
4         Nico Collins         Hou              WR   W 30-6  12.2   

   PASSING YDS  PASSING TD  PASSING INT  RUSHING CAR  RUSHING YDS  ...  \
0          0.0         0.0          0.0         20.0        106.0  ...   
1          0.0         0.0          0.0          0.0          0.0  ...   
2        320.0         4.0          0.0          4.0         17.0  ...   
3          0.0         0.0          0.0          0.0          0.0  ...   
4          0.0         0.0          0.0          0.0          0.0  ...   

   MISC FUML  MISC TD  TOTAL  PASSING COMPLETION  PASSING ATTEMPTS  Opponent  \
0        0.0      0.0   48.7                 0.0               0.0       Ari   
1        0.0      0.0   38.5                 0.0               0.0       Wsh   
2        0.0      0.0   36.5                21.0              25.0       Mia   
3        0.0      0.0   36.0                 0.0               0.0       Mia   
4        0.0      0.0   35.8                 0.0               0.0       Pit   

   Location  rank      DATE  Week  
0      Home   1.0  10-03-23     4  
1      Home   1.0  10-03-23     4  
2      Home   1.0  10-03-23     4  
3      Home   2.0  10-03-23     4  
4      Home   3.0  10-03-23     4  

[5 rows x 26 columns]

In [79]:
dfAll.drop(columns = 'Unnamed: 0', axis = 1, inplace = True)
#using drop function in order to remove unnecessary column

dfAll.head(5)
#previewing dataframe to ensure that it has been dropped

PLAYER NAME PLAYER TEAM PLAYER POSITION   STATUS  PROJ  \
0  Christian McCaffrey          SF              RB  W 35-16  21.1   
1           A.J. Brown         Phi              WR  W 34-31  15.7   
2           Josh Allen         Buf              QB  W 48-20  20.9   
3         Stefon Diggs         Buf              WR  W 48-20  17.8   
4         Nico Collins         Hou              WR   W 30-6  12.2   

   PASSING YDS  PASSING TD  PASSING INT  RUSHING CAR  RUSHING YDS  ...  \
0          0.0         0.0          0.0         20.0        106.0  ...   
1          0.0         0.0          0.0          0.0          0.0  ...   
2        320.0         4.0          0.0          4.0         17.0  ...   
3          0.0         0.0          0.0          0.0          0.0  ...   
4          0.0         0.0          0.0          0.0          0.0  ...   

   MISC FUML  MISC TD  TOTAL  PASSING COMPLETION  PASSING ATTEMPTS  Opponent  \
0        0.0      0.0   48.7                 0.0               0.0       Ari   
1        0.0      0.0   38.5                 0.0               0.0       Wsh   
2        0.0      0.0   36.5                21.0              25.0       Mia   
3        0.0      0.0   36.0                 0.0               0.0       Mia   
4        0.0      0.0   35.8                 0.0               0.0       Pit   

   Location  rank      DATE  Week  
0      Home   1.0  10-03-23     4  
1      Home   1.0  10-03-23     4  
2      Home   1.0  10-03-23     4  
3      Home   2.0  10-03-23     4  
4      Home   3.0  10-03-23     4  

[5 rows x 26 columns]

- Removing unnecessary column ('Unnamed: 0')

In [80]:
print(dfAll[dfAll.duplicated(keep = False)])
print('Duplicated rows: ', dfAll.duplicated().sum())

               PLAYER NAME PLAYER TEAM PLAYER POSITION   STATUS  PROJ  \
3315  Terrace Marshall Jr.         Car              WR  L 24-42   6.3   
3316     Melvin Gordon III         Bal              RB  L 10-17   0.0   
3320       Aidan O'Connell          LV              QB  W 17-13   0.0   
3322        Mitch Trubisky         Pit              QB  W 17-10   0.0   
3365  Terrace Marshall Jr.         Car              WR  L 24-42   6.3   
3366     Melvin Gordon III         Bal              RB  L 10-17   0.0   
3370       Aidan O'Connell          LV              QB  W 17-13   0.0   
3372        Mitch Trubisky         Pit              QB  W 17-10   0.0   

      PASSING YDS  PASSING TD  PASSING INT  RUSHING CAR  RUSHING YDS  ...  \
3315          NaN         NaN          NaN          NaN          NaN  ...   
3316          NaN         NaN          NaN          NaN          NaN  ...   
3320          NaN         NaN          NaN          NaN          NaN  ...   
3322          NaN         NaN     

- Checking for duplicate rows within main dataframe
    - I will be dropping these rows being that there are only 4 duplicates out of 18,090 rows

In [81]:
dfAll = dfAll.drop_duplicates()

- Dropping duplicates

In [87]:
print('Duplicated rows: ', dfAll.duplicated().sum())

Duplicated rows:  0


- Ensuring that the duplicate rows have been dropped

In [93]:
dfAll.isnull().sum()
#checking for null values

PLAYER NAME               0
PLAYER TEAM               0
PLAYER POSITION           0
STATUS                 7067
PROJ                   7097
PASSING YDS           11011
PASSING TD            11011
PASSING INT           11011
RUSHING CAR           11011
RUSHING YDS           11011
RUSHING TD            11011
RECEIVING REC         11011
RECEIVING YDS         11011
RECEIVING TD          11011
RECEIVING TAR         11011
MISC 2PC              11011
MISC FUML             11011
MISC TD               11011
TOTAL                 11060
PASSING COMPLETION    11011
PASSING ATTEMPTS      11011
Opponent               6425
Location                  0
rank                  11060
DATE                      0
Week                      0
dtype: int64

- Columns like player name, player team, and player position have no null values
- Non-Essential columns such as passing yards might represent stats that only apply to certain positions
    - For example, quarterbacks may have passing yard filled, while a running back wont
    - One possible solution is to impute these values with 0
- In order to deal with these nulls, I will handle them on a position by position basis by creating a dataframe for each position

In [83]:
qbs = dfAll[dfAll['PLAYER POSITION'] == 'QB']
#making dataframe with soley quarterback data

qbs.head(5)

PLAYER NAME PLAYER TEAM PLAYER POSITION   STATUS  PROJ  \
2           Josh Allen         Buf              QB  W 48-20  20.9   
7   Anthony Richardson         Ind              QB  L 23-29  20.2   
8        Lamar Jackson         Bal              QB   W 28-3  17.9   
10       Justin Fields         Chi              QB  L 28-31  20.0   
20         Jalen Hurts         Phi              QB  W 34-31  21.7   

    PASSING YDS  PASSING TD  PASSING INT  RUSHING CAR  RUSHING YDS  ...  \
2         320.0         4.0          0.0          4.0         17.0  ...   
7         200.0         2.0          0.0         10.0         56.0  ...   
8         186.0         2.0          0.0          9.0         27.0  ...   
10        335.0         4.0          1.0          4.0         25.0  ...   
20        319.0         2.0          0.0          9.0         34.0  ...   

    MISC FUML  MISC TD  TOTAL  PASSING COMPLETION  PASSING ATTEMPTS  Opponent  \
2         0.0      0.0  36.50                21.0              25.0       Mia   
7         1.0      0.0  29.60                11.0              25.0       LAR   
8         1.0      0.0  28.14                15.0              19.0       Cle   
10        1.0      0.0  27.90                28.0              35.0       Den   
20        0.0      0.0  24.16                25.0              37.0       Wsh   

    Location  rank      DATE  Week  
2       Home   1.0  10-03-23     4  
7       Home   2.0  10-03-23     4  
8       Away   3.0  10-03-23     4  
10      Home   4.0  10-03-23     4  
20      Home   5.0  10-03-23     4  

[5 rows x 26 columns]

In [64]:
wrs = dfAll[dfAll['PLAYER POSITION'] == 'WR']
#making dataframe with soley quarterback data

wrs.head(5)

PLAYER NAME PLAYER TEAM PLAYER POSITION   STATUS  PROJ  PASSING YDS  \
1     A.J. Brown         Phi              WR  W 34-31  15.7          0.0   
3   Stefon Diggs         Buf              WR  W 48-20  17.8          0.0   
4   Nico Collins         Hou              WR   W 30-6  12.2          0.0   
6     Puka Nacua         LAR              WR  W 29-23  14.5          0.0   
13      DJ Moore         Chi              WR  L 28-31  12.3          0.0   

    PASSING TD  PASSING INT  RUSHING CAR  RUSHING YDS  ...  MISC FUML  \
1          0.0          0.0          0.0          0.0  ...        0.0   
3          0.0          0.0          0.0          0.0  ...        0.0   
4          0.0          0.0          0.0          0.0  ...        0.0   
6          0.0          0.0          0.0          0.0  ...        0.0   
13         0.0          0.0          0.0          0.0  ...        0.0   

    MISC TD  TOTAL  PASSING COMPLETION  PASSING ATTEMPTS  Opponent  Location  \
1       0.0   38.5                 0.0               0.0       Wsh      Home   
3       0.0   36.0                 0.0               0.0       Mia      Home   
4       0.0   35.8                 0.0               0.0       Pit      Home   
6       0.0   31.3                 0.0               0.0       Ind      Away   
13      0.0   27.1                 0.0               0.0       Den      Home   

    rank      DATE  Week  
1    1.0  10-03-23     4  
3    2.0  10-03-23     4  
4    3.0  10-03-23     4  
6    4.0  10-03-23     4  
13   5.0  10-03-23     4  

[5 rows x 26 columns]

In [70]:
tes = dfAll[dfAll['PLAYER POSITION'] == 'TE']
#making dataframe with soley quarterback data

tes.head(5)

PLAYER NAME PLAYER TEAM PLAYER POSITION   STATUS  PROJ  PASSING YDS  \
12      Cole Kmet         Chi              TE  L 28-31   7.6          0.0   
17   Mark Andrews         Bal              TE   W 28-3  12.2          0.0   
49    Jonnu Smith         Atl              TE   L 7-23   6.5          0.0   
57  Jake Ferguson         Dal              TE   W 38-3   8.5          0.0   
63  Drew Ogletree         Ind              TE  L 23-29   2.7          0.0   

    PASSING TD  PASSING INT  RUSHING CAR  RUSHING YDS  ...  MISC FUML  \
12         0.0          0.0          1.0          1.0  ...        0.0   
17         0.0          0.0          0.0          0.0  ...        0.0   
49         0.0          0.0          0.0          0.0  ...        0.0   
57         0.0          0.0          0.0          0.0  ...        0.0   
63         0.0          0.0          0.0          0.0  ...        0.0   

    MISC TD  TOTAL  PASSING COMPLETION  PASSING ATTEMPTS  Opponent  Location  \
12      0.0   27.6                 0.0               0.0       Den      Home   
17      0.0   25.0                 0.0               0.0       Cle      Away   
49      0.0   15.5                 0.0               0.0       Jax      Away   
57      0.0   14.7                 0.0               0.0        NE      Home   
63      0.0   13.8                 0.0               0.0       LAR      Home   

    rank      DATE  Week  
12   1.0  10-03-23     4  
17   2.0  10-03-23     4  
49   3.0  10-03-23     4  
57   4.0  10-03-23     4  
63   5.0  10-03-23     4  

[5 rows x 26 columns]

- Creating dataframes for each of position

In [95]:
qbs.isnull().sum()

PLAYER NAME              0
PLAYER TEAM              0
PLAYER POSITION          0
STATUS                 733
PROJ                   738
PASSING YDS           1683
PASSING TD            1683
PASSING INT           1683
RUSHING CAR           1683
RUSHING YDS           1683
RUSHING TD            1683
RECEIVING REC         1683
RECEIVING YDS         1683
RECEIVING TD          1683
RECEIVING TAR         1683
MISC 2PC              1683
MISC FUML             1683
MISC TD               1683
TOTAL                 1685
PASSING COMPLETION    1683
PASSING ATTEMPTS      1683
Opponent               638
Location                 0
rank                  1685
DATE                     0
Week                     0
dtype: int64

- Since metrics like passing stats are essential for quarterbacks, they should ideally have minimal nulls
    - These are likely null due to quarterbacks not playing throwing that game, so I will fill them with zeros as they represent zero activity
    - I will do the same for the other performance related metrics

In [101]:
qbs['PASSING YDS'] = qbs['PASSING YDS'].fillna(0)
qbs['PASSING TD'] = qbs['PASSING TD'].fillna(0)
qbs['PASSING INT'] = qbs['PASSING INT'].fillna(0)
qbs['RUSHING CAR'] = qbs['RUSHING CAR'].fillna(0)
qbs['RUSHING YDS'] = qbs['RUSHING YDS'].fillna(0)
qbs['RUSHING TD'] = qbs['RUSHING TD'].fillna(0)
qbs['RECEIVING REC'] = qbs['RECEIVING REC'].fillna(0)
qbs['RECEIVING YDS'] = qbs['RECEIVING YDS'].fillna(0)
qbs['RECEIVING TD'] = qbs['RECEIVING TD'].fillna(0)
qbs['RECEIVING TAR'] = qbs['RECEIVING TAR'].fillna(0)
qbs['MISC 2PC'] = qbs['MISC 2PC'].fillna(0)
qbs['MISC FUML'] = qbs['MISC FUML'].fillna(0)
qbs['MISC TD'] = qbs['MISC TD'].fillna(0)
qbs['PASSING COMPLETION'] = qbs['PASSING COMPLETION'].fillna(0)
qbs['PASSING ATTEMPTS'] = qbs['PASSING ATTEMPTS'].fillna(0)

/var/folders/q9/jsh43yb95hj4z3v_swlf5_j40000gn/T/ipykernel_4274/83147638.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  qbs['PASSING YDS'] = qbs['PASSING YDS'].fillna(0)
/var/folders/q9/jsh43yb95hj4z3v_swlf5_j40000gn/T/ipykernel_4274/83147638.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  qbs['PASSING TD'] = qbs['PASSING TD'].fillna(0)
/var/folders/q9/jsh43yb95hj4z3v_swlf5_j40000gn/T/ipykernel_4274/83147638.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a Data

***CHECK OUT VARIABLES THAT HAVE NOT BEEN IMPUTED***
- **Opponent and STATUS: These columns may be important for game context, so you could fill null values with placeholders like "Unknown" or "N/A".**
- **PROJ: If you want to keep projections, you could fill nulls with the average projection or a default value of zero if a quarterback doesn’t have a projection.**

In [102]:
qbs.isnull().sum()

PLAYER NAME              0
PLAYER TEAM              0
PLAYER POSITION          0
STATUS                 733
PROJ                   738
PASSING YDS              0
PASSING TD               0
PASSING INT              0
RUSHING CAR              0
RUSHING YDS              0
RUSHING TD               0
RECEIVING REC            0
RECEIVING YDS            0
RECEIVING TD             0
RECEIVING TAR            0
MISC 2PC                 0
MISC FUML                0
MISC TD                  0
TOTAL                 1685
PASSING COMPLETION       0
PASSING ATTEMPTS         0
Opponent               638
Location                 0
rank                  1685
DATE                     0
Week                     0
dtype: int64